playlist_souffle/entities/souffler_util.py

```diff
@@ -0,0 +1,68 @@
+"""Module for souffler entity util functions"""
+
+import random
+import re
+
+SOUFFLE_NAME_RE = r'(?P<playlist_name>.*)\[souffle(?:\^(?P<souffle_degree>\d+))?\]$'
+SOUFFLE_NAME_NO_DEGREE_FMT = '{} [souffle]'
+SOUFFLE_NAME_DEGREE_FMT = '{} [souffle^{}]'
+SOUFFLE_DESCRIPTION_FMT = 'Souffled from "{original_name}" by "{shuffle_by}" at {time_of_souffle}.'
+
+
+def generate_souffle_description(original_name, shuffle_by, time_of_souffle):
+    """Generate a description for a souffled playlist.
+
+    >>> from datetime import datetime as dt
+    >>> time_of_souffle = dt(1968, 11, 22, hour=9, minute=30, second=15)
+    >>> generate_souffle_description('my playlist', 'artist', time_of_souffle)
+    'Souffled from "my playlist" by "artist" at 1968-11-22 09:30:15.'
+    """
+    return SOUFFLE_DESCRIPTION_FMT.format(
+        original_name=original_name,
+        shuffle_by=shuffle_by,
+        time_of_souffle=time_of_souffle
+    )
+
+
+def generate_souffle_name(original_name):
+    """Generate the name of a souffled playlist. On a non-souffled playlist, add [souffle] to the
+    end of playlist name. On a souffled playlist -- a playlist ending in [souffle] -- increment
+    degree, i.e "[souffle^2]".
+
+    >>> generate_souffle_name('my playlist')
+    'my playlist [souffle]'
+
+    >>> generate_souffle_name('my playlist [souffle]')
+    'my playlist [souffle^2]'
+
+    >>> generate_souffle_name('my playlist [souffle^3]')
+    'my playlist [souffle^4]'
+
+    >>> generate_souffle_name('my playlist [souffle^130]')
+    'my playlist [souffle^131]'
+    """
+    match = re.match(SOUFFLE_NAME_RE, original_name)
+
+    if not match:
+        return SOUFFLE_NAME_NO_DEGREE_FMT.format(original_name.rstrip())
+
+    playlist_name = match.group('playlist_name').rstrip()
+    souffle_degree = int(match.group('souffle_degree') or 1)
+
+    return SOUFFLE_NAME_DEGREE_FMT.format(playlist_name, souffle_degree + 1)
+
+
+def souffle_tracks(tracks, collection_tracks_by_track):
```

Comment:
seems like this should be in the `track` entity

Reply:
See comment re: entity awareness of souffleing

Reply:
i'm 50/50 here. souffler could be its own thing, or not. But I don't think you need a util module.

Reply:
This is in souffler_util. Helper functions for souffler but not "public"

Reply:
Oh you said i dont need a util module.

I generally like util modules in python for a bit more encapsulation / clearer public interface for the main module. Uncle bob has some stuff on this re: a "humble object pattern"

Reply:
I think of utils as more generic functions. If a util function relates to a specific entity, then I'd just put it in the module/file of that entity.

I thought humble object was for making function that do IO calls dumb (though it's been a while since I looked that up)